doc/decisions/operation_sequences.md

```diff
@@ -0,0 +1,101 @@
+
+Currently, plugins may (and in the case of notification plugins actually do) assume that a `kdbSet` operation for a certain parent key always directly follows the `kdbGet` operation for that same key. 
+However, this is not enforced in any way within Elektra, nor is it documented that it should.
+
+So in reality, as illustrated by the example, `kdbGet` and `kdbSet` of different parent keys might be interwoven.
+
+```c
+Key * keyA = ...;
+Key * keyB = ...;
+// assume that neither A below B nor B below A
+KeySet * a = ...;
+KeySet * b = ...;
+kdbGet(kdb, a, keyA); // (A)
+kdbGet(kdb, b, keyB); // (B)
+kdbSet(kdb, a, keyA); // (C)
+```
+
+- At point `(A)` the plugins internally store the KeySet for key `a`
+- At point `(B)` the plugins internally override the stored KeySet with the KeySet for key `b`
+- At point `(C)` plugins that do change-tracking will generate a wrong changeset for key `a` if they assume that the previous `kdbGet` operation was also for key `a`
+
+## Constraints
+
+1. As far as possible, we must check for illegal sequences and raise an error when an illegal sequence occurs
+
+## Assumptions
+
+1.
+2.
+3.
+
+## Considered Alternatives
+
+1. Document which sequences are allowed without raising errors in all illegal cases.
+   Least-effort approach, but at least offers some transparency for developers.
+   Developers utilizing Elektra can still ignore it, and may cause problems in certain setups without knowing.
+   
+   (Violates constraint 1)
+
+2. Enforce that the `parentKey` used in `kdbSet` is below the one used in the last `kdbGet`.
+   If this is not the case, `kdbSet` will abort and report an error in `parentKey`.
+   Developers might still wrongly mix the sequences, but they will get an error and have to fix it.
+
+   - Adding a `Key * lastGetParent` to `struct _KDB`
+   - Doing `keyCopy (handle->lastGetParent, parentKey, KEY_CP_NAME)` in `kdbGet`
+   - And checking `keyIsSameOrBelow (handle->lastGetParent, parentKey) == 1` in `kdbSet`
+
+3. Enforce that `ks` used in `kdbSet` is the same as in the last `kdbGet`
+   Similar to (2), but we check for the pointer of the `KeySet` instead of having a copy of the parent key, which saves some memory.
+
+   - Adding a `KeySet * lastGetKs` to `struct _KDB`
+   - Doing `handle->lastGetKs = ks` in `kdbGet`
+   - And checking `lastGetKs == ks` in `kdbSet`
+
+4. Change the API so that a single instance of KDB can only contain a single instance of a `KeySet`.
+   This solution completely eliminates problematic sequences.
+   ```c
+   // same as before but uses handle->data instead of the ks argument
+   int kdbGet (KDB * handle, Key * parentKey);
+   // same as before but uses handle->data instead of the ks argument
+   int kdbSet (KDB * handle, Key * parentKey);
+
+   KeySet * kdbData (KDB * handle) {
+       return handle->data;
+   }
+   ```
+
+5. Allow arbitrary sequences and let each plugin deal with it on a case-by-case basis.
```

Comment:
This description make this option sound far too easy. Please explain more details (deep-dup per parent, in every plugin, etc.). You should also mention that we could keep the deepDup'ed data in `KDB` instead of separately in every plugin and provide an API for plugins to access this.

Reply:
Done